### PR TITLE
Improve the Transform API

### PR DESCRIPTION
**Pull request by JimBobSquarePants**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
This PR dramatically simplifies all affine and projective transforms and fixes some issues we had with affine transforms. 

Two new classes have been introduced into the public domain which allow the composition of matrices:
- `AffineTransformBuilder`
- `ProjectiveTransformBuilder`

All `Transform<TPixel>` methods now utilize these two classes which allows us to **automatically** resize canvases to cater for the transformed vectors.

For example an affine transform can be composed in the following manner:

``` c#
var builder = new AffineTransformBuilder(image.Size())
    .AppendRotateMatrixDegrees(angleDeg)
    .AppendScaleMatrix(new SizeF(sx, sy))
    .AppendTranslationMatrix(new PointF(tx, ty));

image.Mutate(i => i.Transform(builder, KnownResamplers.Bicubic));
```
Both classes also have a `Rectangle` constructor overload that allows the transforming of areas of interest in the source image.

``` c#
var rectangle = new Rectangle(48, 0, 48, 24);

var builder = new AffineTransformBuilder(rectangle)
    .AppendScaleMatrix(new SizeF(2, 1.5F));

image.Mutate(i => i.Transform(builder, KnownResamplers.Spline));
```

Additionally several layers of abstraction have now been removed. And performance of transforms improved by around 10%.

### CODE REVIEW

**Comment by JimBobSquarePants** [comment]
Bump 😉

**Comment by antonfirsov** [comment]
@JimBobSquarePants I will review this during the weekend.

File: src/ImageSharp/Processing/AffineTransformBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        /// Gets the source image size.
+        /// </summary>
+        internal Size Size { get; }
+
+        /// <summary>
+        /// Prepends a centered rotation matrix using the given rotation in degrees.
+        /// </summary>
+        /// <param name="degrees">The amount of rotation, in degrees.</param>
+        /// <returns>The <see cref="AffineTransformBuilder"/>.</returns>
+        public AffineTransformBuilder PrependRotateMatrixDegrees(float degrees)
```
Comment (antonfirsov) [inline src/ImageSharp/Processing/AffineTransformBuilder.cs]: Having the word `Matrix` in the name of these methods seems unnecessarily verbose to me.
We can use it as `PrependRotationDegrees(...)`

We also need a `PrependRotationRadians(...)`.

File: src/ImageSharp/Processing/AffineTransformBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        /// <returns>The <see cref="AffineTransformBuilder"/>.</returns>
+        public AffineTransformBuilder AppendRotateMatrixDegrees(float degrees)
+            => this.AppendMatrix(TransformUtils.CreateRotationMatrixDegrees(degrees, this.Size));
+
+        /// <summary>
+        /// Prepends a scale matrix from the given vector scale.
+        /// </summary>
+        /// <param name="scales">The horizontal and vertical scale.</param>
+        /// <returns>The <see cref="AffineTransformBuilder"/>.</returns>
+        public AffineTransformBuilder PrependScaleMatrix(SizeF scales)
```
Comment (antonfirsov) [inline src/ImageSharp/Processing/AffineTransformBuilder.cs]: `PrependScale(...)`?  ....

File: src/ImageSharp/Processing/Processors/Transforms/AffineTransformProcessor.cs [hunk 3]
```diff
@@ -20,29 +16,35 @@ namespace SixLabors.ImageSharp.Processing.Processors.Transforms
...
-        public AffineTransformProcessor(Matrix3x2 matrix, IResampler sampler, Size targetDimensions)
-            : base(sampler)
+        /// <param name="sourceSize">The source image size.</param>
+        public AffineTransformProcessor(Matrix3x2 matrix, IResampler sampler, Size sourceSize)
         {
+            Guard.NotNull(sampler, nameof(sampler));
+            this.Sampler = sampler;
             this.TransformMatrix = matrix;
-            this.TargetDimensions = targetDimensions;
+            this.TargetDimensions = TransformUtils.GetTransformedSize(sourceSize, matrix);
```
Comment (antonfirsov) [inline src/ImageSharp/Processing/Processors/Transforms/AffineTransformProcessor.cs]: Not sure whether the logic implemented in `GetTransformedSize(...)` fits the needs of all users. Maybe we should expose a `AffineTransformProcessor(matrix, sampler, targetSize)` constructor instead, and deal with the rest in the extension methods.

**Comment by antonfirsov** [comment]
One other suggestion (which is pretty much related to my previous suggestions):

Asking the user to feed the source image size to Transform Builders seems unnecessary to me! The size is available on `IImageProcessingContext<T>` when a transform is being applied. We can prepend it to the matrix created by the Transform Builders. This will simplify the API usage on consumer side, and allow more flexibility.

File: tests/ImageSharp.Tests/Processing/Transforms/ProjectiveTransformBuilderTests.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            b1.AppendMatrix(m4)
+              .AppendTaperMatrix(TaperSide.Left, TaperCorner.LeftOrTop, pi);
+
+            // Backwards
+            b2.PrependTaperMatrix(TaperSide.Left, TaperCorner.LeftOrTop, pi)
+              .PrependMatrix(m4);
+
+            Assert.Equal(b1.BuildMatrix(), b2.BuildMatrix());
+        }
+
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply tests/ImageSharp.Tests/Processing/Transforms/ProjectiveTransformBuilderTests.cs]: Both have 100% coverage as far as I’m aware.

File: tests/ImageSharp.Tests/Processing/Transforms/AffineTransformTests.cs [hunk 5]
```diff
@@ -17,7 +17,8 @@ public class AffineTransformTests
     {
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply tests/ImageSharp.Tests/Processing/Transforms/AffineTransformTests.cs]: Uh oh!

File: tests/ImageSharp.Tests/Processing/Transforms/ProjectiveTransformBuilderTests.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (antonfirsov) [reply tests/ImageSharp.Tests/Processing/Transforms/ProjectiveTransformBuilderTests.cs]: The specific creation methods seemed uncovered for me + we can have a bit more semantical testing.

An idea that's easy to implement: a point transformed by a matrix created with the builder using a series of basic transforms should get to the expected position.

File: src/ImageSharp/Processing/TransformExtensions.cs [hunk 6]
```diff
@@ -14,109 +12,49 @@ namespace SixLabors.ImageSharp.Processing
...
         /// <summary>
-        /// Transforms an image by the given matrix.
+        /// Performs an affine transform of an image.
         /// </summary>
         /// <typeparam name="TPixel">The pixel format.</typeparam>
         /// <param name="source">The image to transform.</param>
-        /// <param name="matrix">The transformation matrix.</param>
+        /// <param name="builder">The affine transform builder.</param>
         /// <returns>The <see cref="Image{TPixel}"/></returns>
-        public static IImageProcessingContext<TPixel> Transform<TPixel>(this IImageProcessingContext<TPixel> source, Matrix3x2 matrix)
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/TransformExtensions.cs]: I dunno, can’t precalculate the correct output canvas then. That’s why size is a requirement of the constructors. The size provided via the context is provided too late.

File: src/ImageSharp/Processing/ProjectiveTransformBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        /// Appends a raw matrix.
+        /// </summary>
+        /// <param name="matrix">The matrix to append.</param>
+        /// <returns>The <see cref="ProjectiveTransformBuilder"/>.</returns>
+        public ProjectiveTransformBuilder AppendMatrix(Matrix4x4 matrix)
+        {
+            this.matrices.Add(matrix);
+            return this;
+        }
+
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/ProjectiveTransformBuilder.cs]: Are they possible with Matrix4x4?

File: src/ImageSharp/Processing/Processors/Transforms/AffineTransformProcessor.cs [hunk 8]
```diff
@@ -64,176 +66,79 @@ protected override Image<TPixel> CreateDestination(Image<TPixel> source, Rectang
...
-                                    // Reverse the premultiplication
-                                    Vector4Utils.UnPremultiply(ref sum);
-                                    dest.FromVector4(sum);
-                                }
+                            for (int x = 0; x < width; x++)
+                            {
+                                // Use the single precision position to calculate correct bounding pixels
+                                // otherwise we get rogue pixels outside of the bounds.
+                                var point = Vector2.Transform(new Vector2(x, y), matrix);
+                                kernel.Convolve(point, x, ref ySpanRef, ref xSpanRef, source.PixelBuffer, vectorSpan);
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/Processors/Transforms/AffineTransformProcessor.cs]: Any optimisation would be welcome. I hate that it’s on the fly

File: src/ImageSharp/Processing/Processors/Transforms/AffineTransformProcessor.cs [hunk 9]
```diff
@@ -64,176 +66,79 @@ protected override Image<TPixel> CreateDestination(Image<TPixel> source, Rectang
...
         {
+            // Handle tranforms that result in output identical to the original.
+            if (this.TransformMatrix.Equals(default) || this.TransformMatrix.Equals(Matrix3x2.Identity))
+            {
+                // The cloned will be blank here copy all the pixel data over
+                source.GetPixelSpan().CopyTo(destination.GetPixelSpan());
+                return;
+            }
+
             int height = this.TargetDimensions.Height;
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/Processors/Transforms/AffineTransformProcessor.cs]: What is? The conditional?

File: src/ImageSharp/Processing/Processors/Transforms/AffineTransformProcessor.cs (hunk 3 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/Processors/Transforms/AffineTransformProcessor.cs]: It’s seems to cover everything I could throw at it, shrinking and growing when need be. There’s actually similar code for handling free transforms in the net framework.

File: src/ImageSharp/Processing/AffineTransformBuilder.cs [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+                matrix *= m;
+            }
+
+            return matrix;
+        }
+
+        /// <summary>
+        /// Removes all matrices from the builder.
+        /// </summary>
+        public void Clear() => this.matrices.Clear();
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/AffineTransformBuilder.cs]: Happy to drop. It was there to allow reuse but yeah, probably not needed.

File: src/ImageSharp/Processing/ProjectiveTransformBuilder.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/ImageSharp/Processing/ProjectiveTransformBuilder.cs]: Yes, Affine is a subset of Projective. Not sure about the details with the `Matrix4x4` class, but I guess you need to use the 3D API-s with `Z=0` for translation and `Z=1` for scale-like stuff.

File: src/ImageSharp/Processing/AffineTransformBuilder.cs [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+        /// <summary>
+        /// Returns the combined matrix.
+        /// </summary>
+        /// <returns>The <see cref="Matrix3x2"/>.</returns>
+        public Matrix3x2 BuildMatrix()
+        {
+            Matrix3x2 matrix = Matrix3x2.Identity;
+
+            // Translate the origin matrix to cater for source rectangle offsets.
+            if (!this.rectangle.Equals(default))
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/AffineTransformBuilder.cs]: Yeah, that’s cleaner.

File: src/ImageSharp/Processing/Processors/Transforms/AffineTransformProcessor.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/ImageSharp/Processing/Processors/Transforms/AffineTransformProcessor.cs]: Sorry, the variable. (At least according to R# suggetion)

File: src/ImageSharp/Processing/AffineTransformBuilder.cs (hunk 2 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/AffineTransformBuilder.cs]: I did experiment with that naming but syntactically it felt incomplete.

File: src/ImageSharp/Processing/TransformExtensions.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/ImageSharp/Processing/TransformExtensions.cs]: You can take the output canvas size from the outside. See my [last comment](https://github.com/SixLabors/ImageSharp/pull/775#issuecomment-441458509).

[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/ImageSharp/Processing/TransformExtensions.cs]: Also related to [this](https://github.com/SixLabors/ImageSharp/pull/775#discussion_r236082294).

[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/TransformExtensions.cs]: Hi would I then fix up the rotation matrix? I need the size to centralise it.

**Comment by antonfirsov** [comment]
@JimBobSquarePants OK, how about this:

I push some more tests + extend the projective API right now, you deal with the rest later?

How about the method namings? Should I change them? Still feel the word `Matrix` is unnecessary method names, and it will become somewhat outdated if we shift to the matrix factory design.

File: src/ImageSharp/Processing/Processors/Transforms/AffineTransformProcessor.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/Processors/Transforms/AffineTransformProcessor.cs]: I’ll double check. I’ve still got R# disabled so my machine doesn’t overheat

File: src/ImageSharp/Processing/AffineTransformBuilder.cs [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+
+namespace SixLabors.ImageSharp.Processing
+{
+    /// <summary>
+    /// A helper class for constructing <see cref="Matrix3x2"/> instances for use in affine transforms.
+    /// </summary>
+    public class AffineTransformBuilder
+    {
+        private readonly List<Matrix3x2> matrices = new List<Matrix3x2>();
+        private Rectangle rectangle;
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/AffineTransformBuilder.cs]: Yeah, I think you’re right. I wrote this on a train with a hangover so might have missed a few things

File: src/ImageSharp/Processing/Processors/Transforms/AffineTransformProcessor.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/ImageSharp/Processing/Processors/Transforms/AffineTransformProcessor.cs]: Yeah it also went crazy for me with the latest updates. My feeling is that something is fundamentally wrong in it when applied to projects with the amount of generics we have.

**Comment by JimBobSquarePants** [comment]
Thanks for the review. Yeah sounds good. Let’s go ahead with naming changes, it is awfully wordy just now.

File: src/ImageSharp/Common/Helpers/ImageMaths.cs [hunk 13]
```diff
@@ -128,6 +130,15 @@ public static int ModuloP2(int x, int m)
             return x & (m - 1);
         }
 
+        /// <summary>
+        /// Converts degrees to radians
+        /// </summary>
+        [MethodImpl(InliningOptions.ShortMethod)]
+        public static float DegreesToRadians(float degrees)
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Common/Helpers/ImageMaths.cs]: `MathFExtensions.DegreeToRadian` in Core provides this already.

File: src/ImageSharp/Processing/ProjectiveTransformBuilder.cs [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+        /// <summary>
+        /// Appends a centered rotation matrix using the given rotation in radians.
+        /// </summary>
+        /// <param name="radians">The amount of rotation, in radians.</param>
+        /// <returns>The <see cref="ProjectiveTransformBuilder"/>.</returns>
+        public ProjectiveTransformBuilder AppendCenteredRotationRadians(float radians)
+        {
+            var m = new Matrix4x4(TransformUtils.CreateRotationMatrixRadians(radians, this.Size));
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Processing/ProjectiveTransformBuilder.cs]: Well this is a pleasant surprise! I didn't know it had a constructor that too a 3x2 matrix :smile:

File: src/ImageSharp/Common/Helpers/ImageMaths.cs (hunk 13 above)
Comment (antonfirsov) [reply src/ImageSharp/Common/Helpers/ImageMaths.cs]: It seemed internal to me.

Comment (JimBobSquarePants) [reply src/ImageSharp/Common/Helpers/ImageMaths.cs]: we use the entire MathF internal class on unsupported target frameworks. Dunno what rules we should apply.

Comment (antonfirsov) [reply src/ImageSharp/Common/Helpers/ImageMaths.cs]: `MathF` might be an exception because it's API is fixed, so we can't break compatibility by accident.

Comment (JimBobSquarePants) [reply src/ImageSharp/Common/Helpers/ImageMaths.cs]: I've updated Core to expose new `GeometryUtilities`

File: src/ImageSharp/Processing/Processors/Transforms/RotateProcessor.cs [hunk 15]
```diff
@@ -34,12 +36,18 @@ public RotateProcessor(float degrees, Size sourceSize)
...
         public RotateProcessor(float degrees, IResampler sampler, Size sourceSize)
-            : base(
+            : this(
                  TransformUtils.CreateRotationMatrixDegrees(degrees, sourceSize),
                  sampler,
                  sourceSize)
             => this.Degrees = degrees;
 
+        // Helper constructor
+        private RotateProcessor(Matrix3x2 rotationMatrix, IResampler sampler, Size sourceSize)
```
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply src/ImageSharp/Processing/Processors/Transforms/RotateProcessor.cs]: I can live with this just now. I don't want to have to do too many calculations outside of these dedicated processors.

**Comment by JimBobSquarePants** [comment]
> And we still may want to find a workaround for the .NET 4.6.2 issue, however it's debug-only. (I think it is Vector2 tricking us again.)

@antonfirsov When you have time could you check this again? When I updated Core etc I think I got an implicit upgrade to System.Numerics.Vectors 4.5.0 so I want to see if there's any changes. Otherwise we may have to do something like the following which is super crappy.

https://github.com/SixLabors/ImageSharp/blob/12a5fdf8aca57ada9eae88481f84900fae7522b8/src/ImageSharp/Processing/Processors/Transforms/ProjectiveTransformProcessor.cs#L145

**Comment by JimBobSquarePants** [comment]
@antonfirsov Aside from the alpha issue in debug I think we're done here.

**Comment by antonfirsov** [comment]
@antonfirsov checking it now. Is the issue reproducible on your PC? (.NET 4.6.2/4.7.2 test execution + Debug)

**Comment by antonfirsov** [comment]
For me it still fails.

We either should do the `Vector2` -> `Vector4` workaround, or leave a note somewhere in the contribution guide that Debug test execution is only supported on .NET Core 2.1.

**Comment by antonfirsov** [comment]
@JimBobSquarePants what is still missing:
- Unit Tests for `TransformUtils`: Should test the created transformations for correctness using sample input points paired with expected transformed destination points. I would say this is critical, but maybe I can live without it, if we can't do it now.
- `Skew` in `ProjectiveTransformBuilder` (Could be added in a follow up PR)
- Unit tests for `AppendSkew`: Similar to rotation tests: weshould test that the transformation being appended/prepended is the same transformation that is produced by `TransformUtils`. (Less critical, could be added in a follow up PR)

**Comment by JimBobSquarePants** [comment]
@antonfirsov I get 283 failed tests running in debug on net472. Even things as simple as `Rgba32` equality. Considering we're using `ushort.Equals` there I don't think they are fixable.

**Comment by JimBobSquarePants** [comment]
> Unit Tests for TransformUtils: Should test the created transformations for correctness using sample input points paired with expected transformed destination points. I would say this is critical, but maybe I can live without it, if we can't do it now.

You're a hard taskmaster you know that right? 😝

File: tests/ImageSharp.Tests/Processing/Transforms/AffineTransformTests.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (JimBobSquarePants) [reply tests/ImageSharp.Tests/Processing/Transforms/AffineTransformTests.cs]: I don't think this error is specific to transforms. This image never runs through that code but is a span copy. I think it's got something to do with `Rgba32`. 

rotate_withrotatetypeenum_testpattern100x50_none

**Comment by antonfirsov** [comment]
Opend #780 to "deal" with the Debug mode issues, we probalby shouldn't bother working around this in product code any longer.

I don't want to be too hard at this PR, up to you to decide. I'm just trying to save our future selves from writing these tests when we touch the code again 😄

**Pull request merged**